### ISSUE TITLE
Audio buffers  

### ISSUE THREAD

**Issue by brummetj** [lukexor/tetanes#175]
# Description

Hey again! I'm still messing around with Tetanes and WebRTC.. fun stuff. I was wondering if you could give some help/advice on how to work with your audio mixer API?

Basically Im trying to encode your audio frame into OPUS, I set the frequency to 48000hz on the audio api, and I'm setting the delta for rate control to `0.005`. This looked to be the delta the https://github.com/libretro/docs/blob/master/archive/ratecontrol.pdf has for 48khz. 

Here is a small simplified gist of what is going on. 

```
// sample_rate * nbr_channels * frame size / 1000. 960

const AUDIO_BUFFER_SIZE = (48000 * 2 * 10 / 1000 ) as usize 

// clock frame (runs the audio.consume() samples fn ) 

self.emulator.clock_frame();

let mut audio_buffer = vec![0f32; AUDIO_BUFFER_SIZE];
self.emulator.audio_callback(&mut audio_buffer);

//opus encoder

let new_buffer = self.encoder.encode_vec_float(audio_buffer, AUDIO_BUFFER_SIZE).unwrap();

//send buffer to webrtc sample track 
```    

This works to the extent where I can hear the audio coming from the buffer, but it is indeed very choppy and crackling sounding. It seems you already are accounting for a ring buffer and apply new audio samples to it. 
 

iv tried to wait for the audio buffer to fill to 20ms (960) before to pass it into the encoder by adding 
```
if self.emulator.audio.len() >= AUDIO_BUFFER_SIZE {
    let mut audio_buffer = vec![0f32; AUDIO_BUFFER_SIZE];
    self.emulator.audio_callback(&mut audio_buffer);

    if let Err(_) = self.audio_tx.send(audio_buffer) {
        log::error!("failed to send audio samples");
    }
}
```

it helps a bit but audio doesn't sound very good still. Am I missing something that you can think of? iv seen some people do some audio buffer stretching, which was my other idea. Maybe it's just a timing issue?

**Comment by lukexor** [comment]
Its been awhile since I've looked at the audio system but I was never able to fully get the dynamic rate control to match performance similar to that paper.

The timing requirements for smooth audio are really tight and it's been something I've been trying to solve. The native emulator works fine on a decent machine, but any lower end machines, web assembly, or other constrained environment results in missed frames resulting in choppy audio as the emulation can't keep up.

I've been meaning to do a major performance pass for awhile now, like trying to put either the emulation or the audio on its own thread, adding SIMD, etc. I just haven't had the time recently to dedicate to it with my son being born and a new job. Hopefully I can get back to it soon. If you have any thoughts on how to help in the meantime let me know!

**Comment by brummetj** [comment]
Congratulations on the newborn and new job!  

I can definitely take a look, I just have a couple questions. 

When you mentioned audio on its own thread, do you mean putting the AudioMixer on its own thread, or splitting the audio out from the bus/cpu ? If you can link me the direction on how to split it apart I can see if I can figure out how to make that happen.

Have you ever looked into using a resampling rust crate like `rubato` ? 

is clock_frame() coupled with audio? meaning that each time you `clock_frame()` you get a new audio sample? or can those samples be grabbed at a different interval.

is it correct the audio samples are produced as a single channel (mono) ? following this assumption does SDL handle all the audio integration with the hardware for you? I think there is some complexity involved with how the audio samples are being produced, using stereo/mono, and getting the 20ms duration of audio. I am fairly new to handling raw audio data.

I usually get around 700-830 samples from your resampling algorithm for a 48khz output frequency. I need a buffer size of exactly 960, is it shorter because the duration of the sound (`control_deck.audio_samples()`) isn't exactly 20ms of sound? any ideas on a good way to get that exact duration?

Thanks for taking the time to read my questions and giving input.

**Comment by lukexor** [comment]
Thanks! Originally I was thinking the audio processing so yes the mixer. However after doing some profiling the mixer isnt actually taking up that much cpu and the SDL2 mixer already is handling the callback on a separate thread.

I have, but currently the mixer is already downsampling the generated audio and applying filters. It might be worth exploring though. The apu runs at 1.789mhz (same as the cpu) which means that's the sampling rate. To downsample to 44100 I'm averaging samples and my attempts to decrease the decimation if there is a dip in framerate havent been successful. The catch is to avoid noticable audio pitch shifting.

Yes audio is lockstep with the cpu which is why emulation frame/audio sync is so hard. One frame is 1/60th a second of audio.

Yes it's mono and yes Ive set up sdl to handle it.

TetaNES is currently configured for 44100hz which is 735 samples every frame. There's a configuration you can set to change it to 48000 which simply changes the sampling rate. But that's still only 800 samples. I'm not sure why you need exactly 20ms of audio for a 16.66ms frame, but the emulation can't provide more without speeding up the emulation. I've seen some emulators mention an audio delay of 20ms but that's different and just means you're about 1 frame of audio behind the emulation vs what you hear, which is imperceptible

The other thing to note is that the default for TetaNES is to be cycle accurate which is required for some quirky games. You can disable it which should give faster framerates.

Basically cycle accurate catches the ppu emulation up on every cpu read or write operation which is a lot of context switching and is probably thrashing the cpu instruction and malemory caches. Not cycle accurate catches the ppu up every full cpu instruction instead.

It could also be possible to get catching the ppu up even less often which some emulators do by trying to predict the next time a cpu read or write might impact the ppu. There are some interesting discussions on the nesdev forums about some of these things

**Comment by brummetj** [comment]
Sorry for the late response on this ticket... hope you have been well. 

I was able to get the audio to work! Since it was written in mono it required me to interleave the samples into 2 channels for stereo and wait until the ring buffer was at 960 or 20ms. Works great now. :) ... the 20ms is needed by the opus encoder. It requires only 20ms of audio for each packet. 

thanks for the information though. I have been playing around with cycle-accurate as well. Nonetheless, great work here. 

I have one more question, what happens when I see this error message? 
```
unexpected PPU memory access at $4542
```
I believe it only happens when using a `clock_frame()`. but when using your `clock_seconds()` with the correct time to run seems to have fixed that issue. 

anyway, feel free to close this out.
